### PR TITLE
New module: win dsc

### PR DESCRIPTION
**Pull request by trondhindenes**
##### SUMMARY
New module: win_dsc

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
win_dsc

##### ANSIBLE VERSION
```
2.4.0
```

##### ADDITIONAL INFORMATION
This module has been sitting in a personal repo, but as its becoming popular it makes sense to add it to Ansible core.
The module allows execution of DSC resources, and takes a dynamic range of parameter depending on the underlying DSC resource.

### CODE REVIEW

**Comment by trondhindenes** [comment]
Looking closer at the tests, seems we don't have any targets with WMF5.0 or 5.1 on them.

**Comment by daBONDi** [comment]
I like it if it gets put in, use it quite often!
Thx for the Work trondhindenes!

**Comment by trondhindenes** [comment]
Thanks @daBONDi , it was about time to get this cleaned up and PRed in ;-)

File: lib/ansible/modules/windows/win_dsc.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dsc.ps1]: We are using normal hashtables these days for the result object.

```powershell
$result = @{
    changed= $false
}
```

File: lib/ansible/modules/windows/win_dsc.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dsc.ps1]: We use this nowadays:
```powershell
$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false
```
It helps if we can keep the same statements/variable naming for the different modules.

File: lib/ansible/modules/windows/win_dsc.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dsc.ps1]: Also use Get-AnsibleParam here.

File: lib/ansible/modules/windows/win_dsc.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dsc.ps1]: These could be simple assignments:
```powershell
$result['resource_name'] = $resourcename
$result['attributes'] = $Attributes
$result['reboot_required'] = $null
```

File: lib/ansible/modules/windows/win_dsc.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dsc.py:42]: I would simply call this `name`, but keep `resource` or `resource_name` as alias. *(Maybe this is not possible because of conflicts in that namespace ?)*

File: lib/ansible/modules/windows/win_dsc.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dsc.py]: The windowsfeature example from the integration tests belongs in this section too. And to be honest, I think we may need a complete section in the documentation related to DSC, real-life examples, etc... This will become the main bulk of what people will be doing with Ansible on Windows going forward.

A full library of common tasks (alternatives to Unix modules) would be very welcome. Something users can easily add examples to, or annotate.

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/windows/win_dsc.py
Comment (trondhindenes) [reply lib/ansible/modules/windows/win_dsc.py]: Yup, I was thinking about that - we'll probably need a separate doc page related to dsc. I'll ask @nitzmahone what he thinks. In the meantime I'll incorporate your suggestions, thanks for reviewing!

Comment (dagwieers) [reply lib/ansible/modules/windows/win_dsc.py]: It's on the agenda for London, so I am confident that once we have decided about the plan, we can collaborate quickly on the implementation. Exciting times !

File: lib/ansible/modules/windows/win_dsc.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dsc.ps1]: That seems weird, having more than one module with the same resource name.
How would you be able to make the distinction ?

Comment (trondhindenes) [reply lib/ansible/modules/windows/win_dsc.ps1]: I know, I was never happy about that part. The idea is that if the "called" DSC resource exists in more than one module, I don't know which one to load. However, I think I'll just default to the latest one, as that's the convention Powershell uses all over. I'll refactor it right now.

**Comment by dagwieers** [comment]
Welcome in the new world order, where we take orders from robots ;-)

File: lib/ansible/modules/windows/win_dsc.py
Comment (trondhindenes) [reply lib/ansible/modules/windows/win_dsc.py:42]: I think "name" is ambigous, as there might also be a "name" param sent into the module (which we rewrite as stated in the docs). I think it's best to avoid "name" altogether on this - and there's a very good chance someone down the line suggests to allow the "name" parameter to be passed in to the underlying dsc resource without rewriting it.

File: lib/ansible/modules/windows/win_dsc.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dsc.py]: This example is not in YAML format.

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/windows/win_dsc.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dsc.ps1]: This could also be a normal assignment:
```yaml
$result[$key.name] = $key.value
```

File: lib/ansible/modules/windows/win_dsc.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dsc.ps1]: Another assignment:
```yaml
$result['DSCAttributes'] = $Attrib
```

File: lib/ansible/modules/windows/win_dsc.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dsc.ps1]: Another one:
```yaml
$result['changed'] = $true
```

File: lib/ansible/modules/windows/win_dsc.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dsc.ps1]: Another one:
```yaml
$result['reboot_required'] = $SetResult.RebootRequired
```

File: lib/ansible/modules/windows/win_dsc.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dsc.ps1:36]: I would simply call this `$resource_name`, so that there is a 1:1 mapping between parameter and variable.
But that's just personal preference.

File: lib/ansible/modules/windows/win_dsc.py
Comment (dagwieers) [reply lib/ansible/modules/windows/win_dsc.py:42]: Ok.

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by dagwieers** [comment]
LGTM I only have cosmetic stuff to add. The module seems to support check-mode and has integration-tests.

I would merge it ASAP so that it allows for broader testing, and providing additional integration-tests and examples as we go forward.

**Comment by dagwieers** [comment]
I was wondering if this module could have universal diff-support included. I don't know if the output leans itself for this, but if it could we should probably add it.

**Comment by trondhindenes** [comment]
Um, I don't know what that is (universal diff-support). I was away from this github repo for a few months, and now everything has changed :-)

**Comment by dagwieers** [comment]
@trondhindenes Well, it's not a generic term, I just meant maybe we can have diff-support in win_dsc that works for all resources equally. This is only possible if the DSC mechanisme provides you *before* and *after* information, or information to what it changed, if there was a change.

File: lib/ansible/modules/windows/win_dsc.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dsc.py]: This can be removed. If a parameter is required, there is no default.

File: lib/ansible/modules/windows/win_dsc.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dsc.py]: If a parameter is not required, we don't have to state that explicitly. It is implicit.
Only when parameters are required, you need to add `required: yes`.

File: test/integration/targets/win_dsc/tasks/main.yml
Comment (dagwieers) [inline test/integration/targets/win_dsc/tasks/main.yml:30]: If you like to avoid this for every task in the integration test, look at this: https://github.com/ansible/ansible/blob/devel/test/integration/targets/win_scheduled_task/tasks/main.yml

We did a few things here:
- A generic test to check capabilities (like you do + version check)
- Only run a specific block based on this test
- Run all tests twice, once in normal mode, then again in check-mode
  - The tests have different *asserts* when the output is expected to be different in check-mode compared to normal mode

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by daBONDi** [comment]
I added some more PropertyType detections into my version of win_dsc maybe you could also add it
Was falling on it when using some 3rd party dsc stuff

```powershell
    # Add Supoort for Int32
    ElseIf ($prop.PropertyType -eq "[Int32]")
    {
        [int]$KeyValue = $attrib.Item($_)
        $config.Property.Add($key,$KeyValue)
    }

    # Add Support for [UInt32]
    ElseIf ($prop.PropertyType -eq "[UInt32]")
    {
        [UInt32]$KeyValue = $attrib.Item($_)
        $config.Property.Add($key,$KeyValue)
    }

    # Add Support for [Int64]
    ElseIf ($prop.PropertyType -eq "[Int64]")
    {
        # Invoke Expression for Support of Strings with String Value: 5GB,5MB,...
        # The Invoke-Expression maybe could be a Security Hole in the Future
        # when a Malcious Runbook executes?
        [Int64]$KeyValue = Invoke-Expression $attrib.Item($_)
        $config.Property.Add($key,$KeyValue)
    }
````

File: lib/ansible/modules/windows/win_dsc.ps1
Comment (daBONDi) [reply lib/ansible/modules/windows/win_dsc.ps1]: Should we not use the defined "Set-Attr" from Ansible, for the Future?
So Ansible Ps Execution Module (Script Includion) also see the changes in the Result Object/Work with it?
Maybe thats another stuff to discuss on Coding standards for Windows Modules :-)

Comment (dagwieers) [reply lib/ansible/modules/windows/win_dsc.ps1]: AFAIU the Set-Attr function existed to have backward compatibility with legacy PS-Object types. But we have been rewriting most modules to use standard hash-tables instead.

I don't understand your second sentence though.

File: test/integration/targets/win_dsc/tasks/main.yml
Comment (trondhindenes) [reply test/integration/targets/win_dsc/tasks/main.yml:30]: I will! I wrote a conditional block for it first, but I must have fat-fingered it. Will check out!

**Comment by trondhindenes** [comment]
Nice @daBONDi we'll take all of the type detections we can get :-).
Not sure if I should leave the PR as-is (with improvements/fixes from @dagwieers's code review comments), and let you have the honor of doing the first feature PR of it afterwards (that way you get credit :-) ). It's up to you!

**Comment by daBONDi** [comment]
@trondhindenes you can add it, you got my permission!, lol :-)

File: lib/ansible/modules/windows/win_dsc.ps1
Comment (daBONDi) [reply lib/ansible/modules/windows/win_dsc.ps1]: I was thinking that "Set-Attr/Get-Attr" want to be like a Wrapper between Ansible Core Powershell Executor and the 3rd Party Plugin, to maybe do some Debugging in the Future, and not a simple Wrapper for working with Objects like it is now. This is also what i mean this should be discussed if it needed for future like Coding Standards and so on...

And idea would be a Standard way in powershell modules for converting python object/json object into a Powershell Object like

Get-Attr -Param "DateAsText" -objecttype [system.datetime] to convert a defined Ansible Yaml Date into an usable .Net DateTime Object

But if this was only for Backwardcompatiblity i'm epic happy to rewrite my plugins internal :-)

thx for clarification!

**Comment by trondhindenes** [comment]
Thanks! Do you have any examples of DSC resources taking parameters of those types? I'm especially wary of the last one, since you're doing an iex to get the value of it.

**Comment by daBONDi** [comment]
The Last one used to Convert a String like "300MB" into an Int64 Representativ, this is why i call Invoke and 
The DSC Resource needs an int64, but i could not figure out specifiy an Int64 with a Number in yaml :-)

So i used the Invoke-Expression and it returns an int64 on GB/MB/...

I needed it on DSC Resource "cFSRMQuotaTemplate" to Create FSRM Quotas.

I think most windows Disk Size stuff needs an int64 because 32bit is not enough :-)
So maybe the Invoke expression is the way to go for the most dsc resources and you can type n1 numbers like "3GB" :-)

**Comment by trondhindenes** [comment]
There are some parse/tryParse methods on the Int64 class in .net, probably safer to do that. I'll have a look.

**Comment by daBONDi** [comment]
yeh but i think we will got a problem when defining more the 2TB Volumes and so on with it, don't think yaml > python will honor ints bigger then 32bit, so you never can define it, or force a String usage :-)

**Comment by trondhindenes** [comment]
@daBONDi I'm gonna hold off on the int64 parser. It might make sense for us to implement a "global" parser for "size-type" strings, I came across this one that Boe write a few years back:
https://gallery.technet.microsoft.com/scriptcenter/262ec604-6b4c-47c8-8940-2b6ce65260ea - something like that. So I'll implement the int32/uint32-support for now. Thanks!

File: test/integration/targets/win_dsc/tasks/main.yml
Comment (trondhindenes) [reply test/integration/targets/win_dsc/tasks/main.yml:30]: Thanks for the example, that looks like a very nice pattern to follow. I think I'm gonna let the test(s) alone for now in this PR, and then get them updated when there's some movement on the ci agent (we need either windows 2012r2 /wmf5.1 or windows 2016 - ideally both)

File: lib/ansible/modules/windows/win_dsc.py
Comment (jborean93) [inline lib/ansible/modules/windows/win_dsc.py:67]: Are you missing the always_run parameter in this or am I missing something?

File: lib/ansible/modules/windows/win_dsc.py
Comment (jborean93) [inline lib/ansible/modules/windows/win_dsc.py:60]: I'm a tiny bit confused your description says parameters for the DSC resource says they should be prefixed with `item_` is that the case or should that be updated. Am I misunderstanding that?

File: test/integration/targets/win_dsc/tasks/main.yml
Comment (jborean93) [inline test/integration/targets/win_dsc/tasks/main.yml]: should be `|success` not `.success`

File: test/integration/targets/win_dsc/tasks/main.yml
Comment (jborean93) [inline test/integration/targets/win_dsc/tasks/main.yml]: should be `|success` not `.success`

File: test/integration/targets/win_dsc/tasks/main.yml
Comment (jborean93) [inline test/integration/targets/win_dsc/tasks/main.yml]: should be `|success` not `.success`

File: test/integration/targets/win_dsc/tasks/main.yml
Comment (jborean93) [inline test/integration/targets/win_dsc/tasks/main.yml]: should be `|success` not `.success`

File: test/integration/targets/win_dsc/tasks/main.yml
Comment (jborean93) [inline test/integration/targets/win_dsc/tasks/main.yml]: should be `|success` not `.success`

File: test/integration/targets/win_dsc/tasks/main.yml
Comment (jborean93) [inline test/integration/targets/win_dsc/tasks/main.yml]: should be `|success` not `.success`

**Review by jborean93 — CHANGES_REQUESTED** [review]
Looks good, fantastic work here. I've just had a comment around the parameters and documentation it doesn't seem too clear to me sorry. Also when running your tests locally on a PS5 box it seems like you need to have `|success` on your checks as `.success` isn't returned in the module output.

I'm the save as Dag where there a few cosmetic stuff but prefer to get this in the wild with more people testing it.

File: lib/ansible/modules/windows/win_dsc.py
Comment (trondhindenes) [reply lib/ansible/modules/windows/win_dsc.py:60]: only the "name" parameter needs to be prefixed. The idea was to avoid confusion with "resource_name" but maybe its a good idea accept 'name' as without any prefix instead.

Comment (jborean93) [reply lib/ansible/modules/windows/win_dsc.py:60]: Ahh that makes sense I've re-read it now and can see what you mean. I would prefer it to be just `name` to keep it consistent with the other stuff but will defer to you for the final decision.

Comment (trondhindenes) [reply lib/ansible/modules/windows/win_dsc.py:60]: Yup agree. I'll get that updated.

Comment (trondhindenes) [reply lib/ansible/modules/windows/win_dsc.py:60]: Fixed the documentation. There was some mismatch between actual module behavior and param documentation. "Name" is very much a valid parameter, if the dsc resource takes a parameter with that name.

File: lib/ansible/modules/windows/win_dsc.py
Comment (trondhindenes) [reply lib/ansible/modules/windows/win_dsc.py:67]: hm, not sure? is that something that needs to be directly supported?

Comment (dagwieers) [reply lib/ansible/modules/windows/win_dsc.py:67]: /me is confused as well :-)

**Comment by dagwieers** [comment]
Personally, I think we should shipit and deal with improving it later.

File: lib/ansible/modules/windows/win_dsc.py
Comment (jborean93) [reply lib/ansible/modules/windows/win_dsc.py:67]: It's probably more the wording of this task, are you saying to run this in ansible_check_mode or some sort of DSC check mode? If it is the former you would need to set `check_mode: True` alongside the task but I don't see a reason why that would need to be documented here.

Comment (dagwieers) [reply lib/ansible/modules/windows/win_dsc.py:67]: Ah, I think it's a copy-and-paste error from the integration tests. Good catch :-)

Comment (trondhindenes) [reply lib/ansible/modules/windows/win_dsc.py:67]: I'm still not quite understanding where/what's missing. There's not documentation to go on, but according to https://groups.google.com/forum/#!topic/ansible-devel/ba4dL-YkMiA, the only thing a Powershell-based module has to do in order to support check mode, is to indicate that using the check_mode parameter when calling "parse-args" - which my module does. Are you saying there's documentation missing?

Roughly how it implements check mode is by calling the test-targetresource DSC function and reporting the "would-be" result, but skipping the "set-targetresource" - which is the function that actually carries out the change. I'm open to improvements, just not able to completely understand what you're saying is missing/wrong.

Comment (jborean93) [reply lib/ansible/modules/windows/win_dsc.py:67]: I probably have this on the wrong line, your name for this task says `name: Invoke DSC with check mode`. If someone comes across this doc and wants to use the module in check mode they might get confused on how to do it. I'm wondering whether you had other intentions for this doc example or whether we should just update the name you have set currently.

Comment (dagwieers) [reply lib/ansible/modules/windows/win_dsc.py:67]: Again, the name for that task was a copy-and-paste from the integration tests.
It would probably better to use, eg.

```yaml
    - name: Enable 'telnet-client' windows feature
```

Comment (trondhindenes) [reply lib/ansible/modules/windows/win_dsc.py:67]: Ah, I understand. Will fix, thanks!

Comment (trondhindenes) [reply lib/ansible/modules/windows/win_dsc.py:67]: So, the idea with the first test is to execute with check mode, and test that the feature is still absent- ie it tests that the module actually supports check mode and doesn't perform any changes.

**Review by dagwieers — APPROVED** [review]

**Review by jborean93 — APPROVED** [review]

**Comment by jborean93** [comment]
shipit

**Comment by dagwieers** [comment]
shipit (our votes don't count though)

**Comment by jborean93** [comment]
Why not, this is a new module and we are Windows contributors?

**Comment by dagwieers** [comment]
I don't think it works that way, you need 2 shipits from maintainers for it to be merged automatically. And for new modules that's not possible. Unless we add ourselves in advance to the MAINTAINERS document ;-)

**Comment by dagwieers** [comment]
bot_status

**Comment by trondhindenes** [comment]
Just reading about the process on https://github.com/ansible/community/blob/master/PR-FLOW.md, but it doesn't completely line up with how the bot labels things. Not sure if I should ping @nitzmahone , just merge it myself (having received two shipits) or something else. Can't wait to squash this madness :-)

**Comment by dagwieers** [comment]
@trondhindenes I've been adding some ideas to the AnsibleFest London EtherPad wrt. contributions and automerging, if you have ideas of improving the workflow as well, share them :-)

2000 open issues and almost 1200 PRs is not looking good :-(

**Comment by nitzmahone** [comment]
Yeah, I think we're at the point that we need to just get it out there for folks to start playing with- I'm sure there will be things to change down the line, but hopefully this can solve some pain for folks that have access to DSC resources that do things Ansible doesn't.

**Comment by trondhindenes** [comment]
sweet, thanks guys! Big step!

**Comment by nitzmahone** [comment]
Thank YOU! Stoked to get it shipped!

**Comment by dagwieers** [comment]
If we finish everything we planned to do at AnsibleFest in London, we can drink MORE beer !

**Comment by jhawkesworth** [comment]
Delighted this is in!  Now need big excuse to start using it.

**Comment by wernerb** [comment]
I am seeing an error with a basic Dsc Resource "File":

```
- name: blabla
  win_dsc5:
    resource_name: File
    ensure: Present
    destinationpath: |-
      C:\example.txt
    Contents: |
      My content
```

Output:

```
Friday 02 June 2017  14:12:45 +0200 (0:00:03.155)       0:00:03.224 ***********
Using module file /Users/werner/projects/nn/aws-platform/ansible/main/playbooks/library/win_dsc5.ps1
<127.0.0.1> ESTABLISH WINRM CONNECTION FOR USER: Administrator on PORT 5986 TO 127.0.0.1
EXEC (via pipeline wrapper)
fatal: [default]: FAILED! => {
    "Attributes": [
        {
            "Key": "ensure",
            "Name": "ensure",
            "Value": "Present"
        },
        {
            "Key": "Contents",
            "Name": "Contents",
            "Value": "My content\n"
        },
        {
            "Key": "destinationpath",
            "Name": "destinationpath",
            "Value": "C:\\example.txt"
        }
    ],
    "Contents": "My content\n",
    "DSCAttributes": {
        "Contents": "My content\n",
        "destinationpath": "C:\\example.txt",
        "ensure": "Present"
    },
    "changed": false,
    "destinationpath": "C:\\example.txt",
    "ensure": "Present",
    "failed": true,
    "resource_name": "File"
}

MSG:

Cannot validate argument on parameter 'ModuleName'. The argument is null or empty. Provide an argument that is not null or empty, and then try the command again.
```

The problem seems to be that there are DscResources that have no module. For File it is empty?

**Comment by trondhindenes** [comment]
File is weird, it has no module. Wont fix. Ill add it the documentation

**Comment by dagwieers** [comment]
I guess that if we're going to add a DSC section to the windows documentation, it's time to start splitting up the Windows docs. It's already one huge document. I'll add it to the WWG agenda.

**Comment by wernerb** [comment]
@trondhindenes I have a patch that allows you to optionally set`module_name` to override. A different case that this also fixes is if you have a resourcename that exists in multiple modules. In both situations you can then supply the correct module name. I will make a pull request in your original repository, and is it welcome here?

An example:

```
- name: Configure IIS/Telnet Client
  win_dsc5:
    resource_name: WindowsFeature
    ensure: Present
    item_name: "Web-server"

- name: blabla
  win_dsc5:
    module_name: PSDesiredStateConfiguration
    resource_name: File
    ensure: Present
    destinationpath: |-
      C:\example.txt
    Contents: |
      My content
```

**Comment by trondhindenes** [comment]
Good point wernerb. Please submit pr to this repo, I won't maintain the "private" version any longer (i'll add some text regarding that in that repo)

**Comment by dagwieers** [comment]
@wernerb From this point forward, PRs have to be created in ansible/ansible.

**Comment by dagwieers** [comment]
Damn you, stale web pages...

**Comment by cwegener** [comment]
I just got here via Trond's message on the old repo. :-)

So glad that this is now merged into ansible. I'm quite a heavy user of this module.

I'll have some time to test all my playbooks with this merged version of the module.

I hope I won't have to open too many issues. ;-) 2000 open issues is quite a lot.

**Comment by dagwieers** [comment]
@cwegener Can we use your expertise for helping with the DSC documentation ?

We would like to have a section, either in the docs or on a wiki, with real-world use-cases, examples and tips-and-tricks to make this attractive to Windows admins. Good quality roles we can point to would be an asset too.

**Comment by cwegener** [comment]
@dagwieers I'd be happy to help with documentation. I'm not sure what the timeframe for the 2.4 release is though (assuming that the DSC documentation should be ready for the release). Depending on what the timeframe is, I can commit to writing up some use cases and examples. I do have a couple of SQL Server playbooks for ansible+dsc that I should be able to share as an example.

**Comment by jborean93** [comment]
@cwegener I'm not sure what @trondhindenes had planned with this but it may be useful to add some of your examples or whatever you think is relevant to https://github.com/ansible/community/wiki/Windows%3A-documentation. That way we can collate it all at the end or add it to the eventual PR that will be there.

**Comment by jhawkesworth** [comment]
@cwegener the timeframe/dates for ansible 2.4 are here: https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_4.rst

**Comment by cwegener** [comment]
@jborean93 Thanks for the link to the wiki. That sounds like a good starting point. I'm pretty new at contributing documentation to ansible. I will need to figure out how to use Github Wikis though. I've never used them before.
@jhawkesworth Looking at the dates for 2.4, I will certainly be able to make some time in my schedule prior to release to document some win_dsc examples. I will start with sharing my ideas/suggestions on the wiki page.

**Comment by trondhindenes** [comment]
I have a ton of examples from our production code aswell, happy to contribute.

**Pull request merged**
